### PR TITLE
[Issue 11036][client] Producer BatcherBuilder configuration property

### LINKED ISSUE #11036: Support configuring the BatcherBuilder through a producer property
**Is your enhancement request related to a problem? Please describe.**
Consumers support `KEY_SHARED` subscriptions. For this to work with producer batching, the producer needs to use a `KEY_BASED` `BatcherBuilder`. The `subscriptionType` of the consumer is currently configurable through properties. The `BatcherBuilder` used by the producer however, is only configurable through code. It would be nice if this could also be done by property.

**Describe the solution you'd like**
Have a configuration property (e.g. `batcherBuilderType`) which supports values `DEFAULT` and `KEY_BASED` for configuring the `BatcherBuilder` used by the producer.

**Describe alternatives you've considered**
None

**Additional context**
Third party products that use pulsar (e.g. debezium) expose the pulsar config through properties. Only being able to configure this in-code is  inconvenient in such cases.

### PR DESCRIPTION
**Pull request by roelrymenants**
Fixes #11036

### Motivation

See #11036 for background on this PR

### Modifications

Added a getter/setter pair for `batcherBuilderType` on the `ProducerConfigurationData` class, which manipulate the `batcherBuilder` attribute, based on an enum. This allows for easy mapping to/from properties and json format.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

- Extended `ConfigurationDataUtilsTest::testLoadProducerConfigurationData`

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no: config property added, not actual cli options)
  - Anything that affects deployment: (no: config property added, but default didn't change)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs: `client-libraries-java.md`)

### CODE REVIEW

File: site2/website/versioned_docs/version-2.6.3/client-libraries-java.md [hunk 1]
```diff
@@ -186,6 +186,7 @@ ProducerCryptoFailureAction|`cryptoFailureAction`|Producer should take action wh
 long|`batchingMaxPublishDelayMicros`|Batching time period of sending messages.|TimeUnit.MILLISECONDS.toMicros(1)
 int|batchingMaxMessages|The maximum number of messages permitted in a batch.|1000
 boolean|`batchingEnabled`|Enable batching of messages. |true
+BatcherBuilderType| `batcherBuilderType`| BatcherBuilder type <br/><br/>Two types are available:<li>Default</li><li>Key_Based</li>|BatcherBuilderType.Default
```
Comment (eolivelli) [inline site2/website/versioned_docs/version-2.6.3/client-libraries-java.md:189]: you are modifying the wrong file, this is for version 2.6.3, please modify the "latest" version of the doc

**Review by eolivelli — CHANGES_REQUESTED** [review]
LGTM
can you please add more unit tests ?
you are modifying an existing test, but we are not covering all the possibilities.

I also left a comment on the docs.
Great to see that you take into account to update the docs !
thank you very much

File: site2/website/versioned_docs/version-2.6.3/client-libraries-java.md (hunk 1 above)
Comment (roelrymenants) [reply site2/website/versioned_docs/version-2.6.3/client-libraries-java.md]: yikes. Reverted and applied to the correct file

**Comment by roelrymenants** [comment]
I added more coverage wrt the batcherBuilderType producer property

**Comment by roelrymenants** [comment]
@eolivelli could you have another look please?

**Comment by codelipenghui** [comment]
The pr had no activity for 30 days, mark with Stale label.

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**